### PR TITLE
Initial work on UntypedObject coercions for binary operators

### PR DESCRIPTION
**Pull request by bryancroteau-MSFT**
Enables coercions from UntypedObject to other types in binary and unary operators. Untyped/Blank combinations often coerce to Number. Includes tests for Blank() for many operators.

### CODE REVIEW

File: src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj [hunk 1]
```diff
@@ -48,6 +48,10 @@
     </Content>
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (bryancroteau-MSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj:51]: Will remove this change

File: src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs [hunk 2]
```diff
@@ -670,18 +671,29 @@ BinderCheckTypeResult ReportInvalidOperation()
             }
         }
 
-        private static BinderCheckTypeResult CheckComparisonArgTypesCore(IErrorContainer errorContainer, TexlNode left, TexlNode right, DType typeLeft, DType typeRight)
+        private static BinderCheckTypeResult CheckComparisonArgTypesCore(IErrorContainer errorContainer, TexlNode binaryOpNode, TexlNode left, TexlNode right, DType typeLeft, DType typeRight)
```
Comment (Shpakh-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs:674]: >TexlNode [](http://example.com/codeflow?start=105&length=8)

Why not this be BinaryOpNode? Left and right can be retrieved from this node.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ParseJson_Arithmetic.txt [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+>> 1=ParseJSON("1")
+true
+
+>> ParseJSON("1")<>1
+false
+
+>> 1<>ParseJSON("1")
+false
+
+>> ParseJSON("1")=Blank()
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ParseJson_Arithmetic.txt:128]: >ParseJSON("1")=Blank() [](http://example.com/codeflow?start=3&length=22)

Some additional tests (sorry if they are defined further down):

    >> ParseJSON("0") = Blank()
    false

    >> ParseJSON("null") = Blank()
    true

Same with the order swapped

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/CustomRecordTypeWithUndefinedFields.cs [hunk 4]
```diff
@@ -79,7 +79,7 @@ public async Task TestOriginalRecordType(string expr, object expected, bool hasE
         [InlineData("IsBlank(obj.missing)", true, false)] // fields not defined in RecordType is treated as Blank.
         [InlineData("IsBlank(obj.missing.missing)", true, false)]
         [InlineData("IsBlank(obj.prop_not_defined_in_type)", true, false)] // fields not defined in RecordType is always Blank, regardless of whether it is defined in actual data.
-        [InlineData("IsError(obj.missing + 1)", null, true)] // type checking fails when operating on undefined field.
+        [InlineData("IsError(obj.missing + 1)", false, false)] // type checking fails when operating on undefined field.
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/CustomRecordTypeWithUndefinedFields.cs:82]: >/ type checking fails when operating on undefined field [](http://example.com/codeflow?start=64&length=55)

Update comment

File: src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs [hunk 5]
```diff
@@ -769,6 +801,20 @@ private static BinderCheckTypeResult CheckEqualArgTypesCore(IErrorContainer erro
                 return new BinderCheckTypeResult();
             }
 
+            if (typeLeft.IsUntypedObject && typeRight.IsUntypedObject)
+            {
+                if (typeLeft.IsUntypedObject && typeRight.IsUntypedObject)
```
Comment (Shpakh-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs:806]: >typeLeft.IsUntypedObject && typeRight.IsUntypedObject) [](http://example.com/codeflow?start=20&length=54)

Why check again?

File: src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs [hunk 6]
```diff
@@ -635,7 +636,7 @@ BinderCheckTypeResult ReportInvalidOperation()
 
                         default:
                             // Time + number = Time
-                            var resRight = CheckTypeCore(errorContainer, node.Right, rightType, DType.Number, /* coerced: */ DType.String, DType.Boolean);
+                            var resRight = CheckTypeCore(errorContainer, node.Right, rightType, DType.Number, /* coerced: */ DType.String, DType.Boolean, DType.UntypedObject);
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs:639]: >, DType.UntypedObject [](http://example.com/codeflow?start=152&length=21)

We should have tests for those scenarios:

- Time + untyped (and vice-versa)
- Date + untyped (and vice-versa)
- DateTime + untyped (and vice-versa)

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/BasicCoercion.txt [hunk 7]
```diff
@@ -46,7 +46,7 @@ true
 
 // Erroneous Date -> Bool coercion
 >> Date(2000,1,2) && 1
-Errors: Error 0-14: Invalid argument type. Expecting one of the following: Boolean, Number, Text, OptionSetValue.
+Errors: Error 0-14: Invalid argument type. Expecting one of the following: Boolean, Number, Text, OptionSetValue, UntypedObject.
```
Comment (Shpakh-MSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/BasicCoercion.txt:49]: >UntypedObject [](http://example.com/codeflow?start=114&length=13)

Just noticed that, in public documentation, we reference this as 'Untyped object'.

**Comment by MikeStall** [comment]
What exactly are the coercion rules? 

Are the any ambiguities, especially since + can be both number and date, how do we statically resolve?

ParseJson() + 5  // is this NumberAdd, or DateAdd?

**Comment by Shpakh-MSFT** [comment]
Single type operator coerce support is limited to Number, Boolean, Text types. So I assume that this would be NumberAdd. DateAdd would require explicit cast.

---
In reply to: [1445700198](https://github.com/microsoft/Power-Fx/pull/1116#issuecomment-1445700198) [](http://example.com/codeflow?ancestors=1445700198,1445700198)

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs (hunk 5 above)
Comment (bryancroteau-MSFT) [reply src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs:806]: Copy paste error.

File: src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs (hunk 2 above)
Comment (bryancroteau-MSFT) [reply src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs:674]: In other places in the file we use leftNode.Parent, so I'll do that.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ParseJson_Arithmetic.txt (hunk 3 above)
Comment (bryancroteau-MSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ParseJson_Arithmetic.txt:128]: I added these tests. It required adding a new operator implementation specifically for blank+untyped for the eq and neq operators.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/CustomRecordTypeWithUndefinedFields.cs (hunk 4 above)
Comment (bryancroteau-MSFT) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/CustomRecordTypeWithUndefinedFields.cs]: Done

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/BasicCoercion.txt (hunk 7 above)
Comment (bryancroteau-MSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/BasicCoercion.txt:49]: In all our binder error messages we use the DType enum name

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryOperators.cs [hunk 8]
```diff
@@ -466,13 +485,55 @@ private static BooleanValue AreEqual(IRContext irContext, FormulaValue[] args)
...
+        {
+            var arg1 = args[0];
+            var arg2 = args[1];
+
+            if (arg1 is BlankValue && arg2 is BlankValue)
+            {
+                return new BooleanValue(irContext, false);
+            }
+
+            Contracts.Assert(arg1 is UntypedObjectValue ^ arg2 is UntypedObjectValue, "UO = UO is undefined");
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryOperators.cs:526]: >UO = UO [](http://example.com/codeflow?start=87&length=7)

`UO <> UO`

Comment (bryancroteau-MSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryOperators.cs]: Fixed

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by CarlosFigueiraMSFT — DISMISSED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by CarlosFigueiraMSFT — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**